### PR TITLE
Adding updating nodehandle subscribe function.

### PR DESCRIPTION
**Pull request by mikeodr**
Nodehandle subscribe function did not support non-static
function for callbacks properly.

Also added related test.

### CODE REVIEW

**Comment by mikeodr** [comment]
@mikepurvis FYI.

File: rosserial_test/src/publish_subscribe.cpp [hunk 1]
```diff
@@ -47,6 +47,28 @@ static void rosserial_string_cb(const rosserial::std_msgs::String& msg)
...
+    last_string_ = std::string(msg.data);
+  }
+
+  int getCount() { return rosserial_string_cb_count_; }
+  std::string getLastString() { return last_string_; }
+
+private:
+  int rosserial_string_cb_count_;
+  std::string last_string_;
+};
```
Comment (mikepurvis) [inline rosserial_test/src/publish_subscribe.cpp]: May not be necessary, if we can just reuse the one from here: https://github.com/ros-drivers/rosserial/blob/jade-devel/rosserial_test/include/rosserial_test/helpers.h#L3

File: rosserial_client/src/ros_lib/ros/node_handle.h [hunk 2]
```diff
@@ -348,8 +348,8 @@ namespace ros {
       }
 
       /* Register a new subscriber */
-      template<typename MsgT>
-      bool subscribe(Subscriber< MsgT> & s){
+      template<typename Subscriber>
+      bool subscribe(Subscriber& s){
```
Comment (mikepurvis) [inline rosserial_client/src/ros_lib/ros/node_handle.h]: I know I suggested this, but given the other types which exist in this scope, `SubscriberT` might be a less confusing name.

File: rosserial_client/src/ros_lib/ros/node_handle.h [hunk 3]
```diff
@@ -348,8 +348,8 @@ namespace ros {
...
         for(int i = 0; i < MAX_SUBSCRIBERS; i++){
           if(subscribers[i] == 0){ // empty slot
             subscribers[i] = (Subscriber_*) &s;
```
Comment (mikepurvis) [inline rosserial_client/src/ros_lib/ros/node_handle.h]: While you're here anyway, please convert this to a proper C++ `static_cast<Subscriber_*>(&s)`.

File: rosserial_test/src/publish_subscribe.cpp [hunk 4]
```diff
@@ -71,6 +93,32 @@ TEST_F(SingleClientFixture, single_subscribe) {
   EXPECT_EQ(string_msg.data, last_string);
 }
 
+/**
+ * Single message sent from a roscpp Publisher, received
+ * by a rosserial client subscriber.
+ */
+TEST_F(SingleClientFixture, single_subscribe_non_static) {
```
Comment (mikepurvis) [inline rosserial_test/src/publish_subscribe.cpp]: Consider `single_subscribe_member_function` as the name of this.

File: rosserial_client/src/ros_lib/ros/node_handle.h (hunk 2 above)
Comment (mikeodr) [reply rosserial_client/src/ros_lib/ros/node_handle.h]: Got it, done.

File: rosserial_client/src/ros_lib/ros/node_handle.h (hunk 3 above)
Comment (mikeodr) [reply rosserial_client/src/ros_lib/ros/node_handle.h]: Done.

File: rosserial_test/src/publish_subscribe.cpp (hunk 4 above)
Comment (mikeodr) [reply rosserial_test/src/publish_subscribe.cpp]: Done.

**Comment by mikepurvis** [comment]
Thanks, this was an oversight from the previous addition.

**Pull request merged**